src/nepiada/env/nepiada.py

```diff
@@ -191,16 +199,52 @@ def step(self, actions):
 
         # The observation returned in 'step' function is the current position of all agents
         # Note this is different from observation graph!
+        # Each agent is able to directly attain the position of every agent in it's observation graph
         self.observations = {agent: None for agent in self.agents}
         for agent_name in self.agents:
-            agent = self.world.get_agent(agent_name)
-            self.observations[agent_name] = [agent.p_pos[0], agent.p_pos[1]] 
-
-        # Info will be used to pass information about comm and obs graphs
+            observation = {}
+            for observed_agent_name in self.agents:
+                observed_agent = self.world.get_agent(observed_agent_name)
+                if observed_agent_name in self.world.graph.obs[agent_name]:
+                    observation[observed_agent_name] = (observed_agent.p_pos[0], observed_agent.p_pos[1])
+                else:
+                    observation[observed_agent_name] = None # Cannot be observed
+            self.observations[agent_name] = observation
+
+        
+        # First pass observed beliefs
+        for agent_name in self.agents:
+            beliefs = self.world.get_agent(agent_name).beliefs
+            observation = self.observations[agent_name]
+            for target_agent_name in self.agents:
+                if observation[target_agent_name]: # Can directly see
+                    beliefs[target_agent_name] = observation[target_agent_name]
+                else: # Must estimate where the agent is via communication
+                    beliefs[target_agent_name] = None
+
+        # Second pass communicated beliefs
+        for agent_name in self.agents:
+            beliefs = self.world.get_agent(agent_name).beliefs
+            observation = self.observations[agent_name]
+            for target_agent_name in self.agents:
+                if not observation[target_agent_name]: # Must estimate where the agent is via communication
+                    val_x, val_y = [],[]
+                    for helpful_agent in self.world.graph.comm[agent_name]:
+                        helpful_beliefs = copy.deepcopy(self.world.get_agent(helpful_agent).beliefs)
+                        #TODO add noise from our noise class if helpful_agent is adversarial  
```

Comment:
This should probably be included as part of this PR.

Reply:
I agree a good rule of thumb is to include a checklist in your PR of what has been addressed and what needs to be addressed:

- [x] This was changed
- [ ] This needs to change

These are minor corrections for now but will greatly simplify our collaboration when the code gets big and complex